sklearn/preprocessing/imputation.py

```diff
@@ -309,36 +338,57 @@ def _dense_fit(self, X, strategy, missing_values, axis):
                 row = row[row_mask]
                 most_frequent[i] = _most_frequent(row, np.nan, 0)
 
-            return most_frequent
+            statistics = most_frequent
+
+        if axis == 0:
+            return statistics, feat_with_missing
+        else:
+            return statistics
 
-    def _sparse_transform(self, X, valid_stats, valid_idx):
+    def _sparse_transform(self, X, statistics):
         """transformer on sparse data."""
+        imputer_mask = _get_mask(X.data, self.missing_values)
+        mask_matrix = X.__class__((imputer_mask, X.indices.copy(),
+                                   X.indptr.copy()), shape=X.shape,
+                                  dtype=X.dtype)
+        X, valid_stats = self._handle_all_missing(X, statistics)
+
+        if self.axis == 0:
+            all_missing = np.where(mask_matrix.sum(axis=0) == X.shape[0])[0]
+            mask_matrix = mask_matrix[:, self.imputed_features_]
+            if not np.array_equal(all_missing, self.all_missing_):
+                features = np.setdiff1d(self.all_missing_, all_missing)
```

Comment:
So: use standard debugging techniques. It says they're different numbers of dimensions, so one of these is 2d (perhaps a `numpy.matrix`?). Where is each of them set?